builtins/web/blob.cpp

```diff
@@ -0,0 +1,703 @@
+      auto len = chars.len;
+      blob->insert(blob->end(), src, src + len);
+    }
+    return true;
+  }
+
+  // FALLBACK: if we ever get here convert, to string and call append again
+  auto str = JS::ToString(cx, val);
+  if (!str) {
+    return false;
+  }
+
+  RootedValue str_val(cx, JS::StringValue(str));
+  return append_value(cx, self, str_val);
+}
+
+bool Blob::init_blob_parts(JSContext *cx, HandleObject self, HandleValue value) {
+  JS::ForOfIterator it(cx);
+  if (!it.init(value, JS::ForOfIterator::AllowNonIterable)) {
+    return false;
+  }
+
+  bool is_iterable = value.isObject() && it.valueIsIterable();
+
+  if (is_iterable) {
+    // if the object is an iterable, walk over its elements...
+    JS::Rooted<JS::Value> item(cx);
+    while (true) {
+      bool done;
+
+      if (!it.next(&item, &done)) {
+        return false;
+      }
+      if (done) {
+        break;
+      }
+
+      if (!append_value(cx, self, item)) {
+        return false;
+      }
+    }
+
+    return true;
+  } else {
+    // non-objects are not allowed for the blobParts
+    return api::throw_error(cx, api::Errors::TypeError, "Blob.constructor", "blobParts", "be an object");
+  }
+}
+
+bool Blob::init_options(JSContext *cx, HandleObject self, HandleValue initv) {
+  JS::RootedValue init_val(cx, initv);
+
+  if (!init_val.isObject()) {
+    return api::throw_error(cx, api::Errors::TypeError, "Blob.constructor", "options", "be an object");
+  }
+
+  // `options` is an object which may specify any of the properties:
+  // - `type`: the MIME type of the data that will be stored into the blob,
+  // - `endings`: how to interpret newline characters (\n) within the contents.
+  JS::RootedObject opts(cx, init_val.toObjectOrNull());
+  bool has_endings, has_type;
+
+  if (!JS_HasProperty(cx, opts, "endings", &has_endings) ||
+      !JS_HasProperty(cx, opts, "type", &has_type)) {
+    return false;
+  }
+
+  if (!has_type && !has_endings) {
+    // Use defaults
+    return true;
+  }
+
+  if (has_endings) {
+    JS::RootedValue endings_val(cx);
+    bool is_transparent, is_native;
+    if (!JS_GetProperty(cx, opts, "endings", &endings_val)) {
+      return false;
+    }
+
+    auto endings_str = JS::ToString(cx, endings_val);
+    if (!JS_StringEqualsLiteral(cx, endings_str, "transparent", &is_transparent) ||
+        !JS_StringEqualsLiteral(cx, endings_str, "native", &is_native)) {
+      return false;
+    }
+
+    if (is_transparent || is_native) {
+      auto endings = is_native ? LineEndings::Native : LineEndings::Transparent;
+      SetReservedSlot(self, static_cast<uint32_t>(Slots::Endings), JS::Int32Value(endings));
+    }
+  }
+
+  if (has_type) {
+    JS::RootedValue type(cx);
+    if (!JS_GetProperty(cx, opts, "type", &type)) {
+      return false;
+    }
+
+    auto type_str = normalize_type(cx, type);
+    if (!type_str) {
+      return false;
+    }
+    SetReservedSlot(self, static_cast<uint32_t>(Slots::Type), PrivateValue(type_str));
```

Comment:
You should store `type_str` as a string instead of a private value to ensure that the GC knows about it. Changing `PrivateValue` to `StringValue` here and adapting the other sites accessing `Slots::Type` should be all that's needed.